signature-aggregator/api/api.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package api
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/signature-aggregator/aggregator"
+	"github.com/ava-labs/awm-relayer/types"
+	"github.com/ava-labs/awm-relayer/utils"
+	"go.uber.org/zap"
+)
+
+const (
+	RawMessageAPIPath = "/aggregate-signatures/by-raw-message"
+	defaultQuorumNum  = 67
+)
+
+// Defines a request interface for signature aggregation for a raw unsigned message.
+// Currently a copy of the `ManualWarpMessageRequest` struct in relay_message.go
+type AggregateSignaturesByRawMsgRequest struct {
+	// Required. hex-encoded message, optionally prefixed with "0x".
+	UnsignedMessage string `json:"unsigned-message"`
+	// Optional hex or cb58 encoded signing subnet ID. If omitted will default to the subnetID of the source BlockChain
+	SigningSubnetID string `json:"signing-subnet-id"`
+	// Optional. Integer from 0 to 100 representing the percentage of the quorum that is required to sign the message
+	// defaults to 67 if omitted.
+	QuorumNum uint64 `json:"quorum-num"`
+}
+
+type AggregateSignaturesResponse struct {
+	// hex encoding of the signature
+	SignedMessage string `json:"signed-message"`
+}
+
+func HandleAggregateSignaturesByRawMsgRequest(
+	logger logging.Logger,
+	signatureAggregator *aggregator.SignatureAggregator,
+) {
+	http.Handle(RawMessageAPIPath, signatureAggregationAPIHandler(logger, signatureAggregator))
+}
+
+func writeJsonError(
+	logger logging.Logger,
+	w http.ResponseWriter,
+	errorMsg string,
+) {
+	resp, err := json.Marshal(struct{ error string }{error: errorMsg})
+	if err != nil {
+		msg := "Error marshalling JSON error response"
+		logger.Error(msg, zap.Error(err))
+		resp = []byte(msg)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+
+	w.Write(resp)
+	if err != nil {
+		logger.Error("Error writing error response", zap.Error(err))
+	}
+}
+
+func signatureAggregationAPIHandler(logger logging.Logger, aggregator *aggregator.SignatureAggregator) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
For readability, can we define the handler function as a standalong func, and then pass it to `HandlerFunc` by reference?

Reply:
I'm not actually sure what do you mean here. The `HandlerFunc` needs it's signature to be `w http.ResponseWriter, r *http.Request`and the anonymous function inside that we are returning needs access to `logger` and `aggregator` as well.  Is there an example doing what you mean?

Reply:
Sorry I was confusing this with how we've set it up in other places using a `struct` to hold the `logger`/`aggregator` as members, and named handler `func`s. I don't think it's worth doing that here now though.